File: 프로그래머스/3/150365. 미로 탈출 명령어/미로 탈출 명령어.cpp
#include <bits/stdc++.h>

#define MAX 51
using namespace std;
// d l r u
int dx[] = {1, 0, 0, -1};
int dy[] = {0, -1, 1, 0};
char dc[] = {'d', 'l', 'r', 'u'};
int N, M, X, Y, R, C, K;
string answer = "";
string dp[MAX][MAX];

void f(int n, int m, int x, int y, int r, int c, int k) {
    N = n;
    M = m;
    X = x;
    Y = y;
    R = r;
    C = c;
    K = k;
}
bool isImpossible(int x, int y, int remainK) {
    int pathLen = abs(x - R) + abs(y - C);
    if(abs(pathLen - remainK) % 2)
        return true;
    if(pathLen > remainK)
        return true;
    return false;
}
bool cmpString(string A, string B){
    int len = min(A.length(), B.length());
    
    for(int i = 0 ; i < len ; i++) {
        if(A[i] < B[i])
            return true;
    }
    
    return false;
}
void dfs(int curX, int curY) {
    int remainK = K - dp[curX][curY].length();
    // if(remainK <= 0)
    //     return;
    // if(curX == R && curY == C) {
    //     return;
    // }
    if(isImpossible(curX, curY, remainK)) {
        return;
    }
    
    for(int i = 0 ; i < 4 ; i++) {
        int ny = curY + dy[i];
        int nx = curX + dx[i];
        if(ny <= 0 || nx <= 0 || ny > M || nx > N)
            continue;
        string tmp = dp[curX][curY] + dc[i];
        if(tmp.length() > K)
            continue;
        if(cmpString(dp[nx][ny], tmp))
            continue;
        // if(nx == R && ny == C && remainK != 1)
        //     continue;
        dp[nx][ny] = tmp;
        
        dfs(nx, ny);
    }
}
string solution(int n, int m, int x, int y, int r, int c, int k) {
    f(n, m, x, y, r, c, k);
    if(isImpossible(x, y, k))
        return "impossible";
    
    dfs(x, y);
    answer = dp[r][c];
    return answer;
}

File: 프로그래머스/3/150365. 미로 탈출 명령어/미로 탈출 명령어.js
let N, M, R, C;
let answer = 'impossible';

const dx = [1, 0, 0, -1];
const dy = [0, -1, 1, 0];
const dc = ['d', 'l', 'r', 'u'];

function solution(n, m, x, y, r, c, k) {
    N = n;
    M = m;
    R = r;
    C = c;
    
    if(getLength(y, x) % 2 !== k % 2) {
        return answer;
    }
    dfs(y, x, k, '');
    return answer;
}

function dfs(curY, curX, curK, curRoute) {
    if(answer !== 'impossible')
        return;
    if(getLength(curY, curX) > curK) {
        return;
    }
    if(curK === 0) {
        if(curX === R && curY === C) {
            answer = curRoute;
        }
        
        return;
    }
    
    for(let i = 0 ; i < 4 ; i++) {
        const nextY = curY + dy[i];
        const nextX = curX + dx[i];
        if(nextY < 1 || nextY < 1 || nextX > N || nextY > M) {
            continue;
        }
        
        dfs(nextY, nextX, curK - 1 , curRoute + dc[i]);
    }
}

function getLength(curY, curX) {
    return Math.abs(curX - R) + Math.abs(curY - C);
}